ta/gatekeeper/gatekeeper_ta.c

```diff
@@ -0,0 +1,783 @@
+	 * | current_password_handle_length | 4                               |
+	 * | current_password_handle        | #current_password_handle_length |
+	 * +--------------------------------+---------------------------------+
+	 */
+	uint32_t uid;
+	uint32_t desired_password_length;
+	const uint8_t *desired_password;
+	uint32_t current_password_length;
+	const uint8_t *current_password;
+	uint32_t current_password_handle_length;
+	const uint8_t *current_password_handle;
+
+	const uint8_t *request = (const uint8_t *)params[0].memref.buffer;
+	const uint8_t *i_req = request;
+	struct failure_record record;
+
+	/*
+	 * Enroll response layout
+	 * +--------------------------------+---------------------------------+
+	 * | Name                           | Number of bytes                 |
+	 * +--------------------------------+---------------------------------+
+	 * | error                          | 4                               |
+	 * +--------------------------------+---------------------------------+
+	 * | timeout                        | 4                               |
+	 * +------------------------------ OR --------------------------------+
+	 * | password_handle_length         | 4                               |
+	 * | password_handle                | #password_handle_length         |
+	 * +--------------------------------+---------------------------------+
+	 */
+	uint32_t error = GK_ERROR_NONE;
+	uint32_t timeout = 0;
+	struct password_handle password_handle;
+
+	uint8_t *response = params[1].memref.buffer;
+	uint8_t *i_resp = response;
+
+	const uint32_t max_response_size = sizeof(uint32_t) +
+		sizeof(uint32_t) +
+		sizeof(struct password_handle);
+
+	secure_id_t user_id = 0;
+	uint64_t flags = 0;
+	salt_t salt;
+
+	deserialize_int(&i_req, &uid);
+	deserialize_blob(&i_req, &desired_password, &desired_password_length);
+	deserialize_blob(&i_req, &current_password, &current_password_length);
+	deserialize_blob(&i_req, &current_password_handle,
+			&current_password_handle_length);
+
+	/* Check request buffer size */
+	if (get_size(request, i_req) > params[0].memref.size) {
+		EMSG("Wrong request buffer size");
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	/* Check response buffer size */
+	if (max_response_size > params[1].memref.size) {
+		EMSG("Wrong response buffer size");
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	/* Check password handle length */
+	if (current_password_handle_length != 0 &&
+	    current_password_handle_length != sizeof(struct password_handle)) {
+		EMSG("Wrong password handle size");
+		res = TEE_ERROR_BAD_PARAMETERS;
+		goto exit;
+	}
+
+	if (!current_password_handle_length) {
+		/*
+		 * Password handle does not match what is stored, generate new
+		 * secure user_id
+		 */
+		TEE_GenerateRandom(&user_id, sizeof(user_id));
+	} else {
+		uint64_t timestamp;
+		bool throttle;
+
+		struct password_handle *pw_handle = (struct password_handle *)
+						    current_password_handle;
```

Comment:
Looks like client is manipulating TA memory pointers. This should crosscheck the memory reference is a valid registered `struct password_handle` instance the TA published to the client.

Reply:
Frankly, there is no need in pointer usage here nor in `TA_DoVerify` function , this existed in the initial implementation.
I'll fix this